### PR TITLE
Annotation service and model

### PR DESCRIPTION
**Pull request by niio972**
Add annotation concept functionnality defined as Web Annotation Ontology in the web application client. 
References https://github.com/OpenSILEX/phis-ws/issues/64, references #18 .

### CODE REVIEW

File: views/layouts/main.php [hunk 1]
```diff
@@ -19,6 +20,7 @@
 use yii\bootstrap\NavBar;
 use yii\widgets\Breadcrumbs;
 use app\assets\AppAsset;
+use rmrevin\yii\fontawesome\FAS;
```
Comment (Gouga34) [inline views/layouts/main.php]: why did you add this import ?

Comment (niio972) [reply views/layouts/main.php]: In order to add the annotation flag I had to add FontAwesome icons. It is the css dependency.

File: models/yiiModels/YiiVocabularyModel.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        foreach ($array as $namespace) {
+            $this->namespaces[$namespace->prefix] = $namespace->namespace;
+        }
+    }
+
+    /**
+     * Return namespaces list
+     * @param string $sessionToken
+     * @param string $uri
+     * @return mixed l'objet s'il existe, un message sinon
```
[the remark this answers is not in the document]
Comment (niio972) [reply models/yiiModels/YiiVocabularyModel.php]: Done

File: models/yiiModels/YiiUserModel.php [hunk 3]
```diff
@@ -173,7 +199,34 @@ public function findByEmail($sessionToken, $email) {
             return $requestRes;
         }
     }
-    
+
+     /**
+     * Return user searched by uri
+     * @param string $sessionToken
+     * @param string $uri
+     * @return mixed l'objet s'il existe, un message sinon
```
[the remark this answers is not in the document]
Comment (niio972) [reply models/yiiModels/YiiUserModel.php]: Done

File: models/yiiModels/YiiAnnotationModel.php [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            $elementForWebService[YiiAnnotationModel::COMMENTS] = $this->comments;
+        }
+        return $elementForWebService;
+    }
+
+    /**
+     * Find an annotation by this uri
+     * @param string $sessionToken
+     * @param string $uri
+     * @return mixed l'objet s'il existe, un message sinon
```
[the remark this answers is not in the document]
Comment (niio972) [reply models/yiiModels/YiiAnnotationModel.php]: Done

File: models/yiiModels/YiiAnnotationModel.php [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    protected function arrayToAttributes($array) {
+        $this->uri = $array[YiiAnnotationModel::URI];
+        $this->creator = $array[YiiAnnotationModel::CREATOR];
+        $this->comments = $array[YiiAnnotationModel::COMMENTS];
+        $this->motivatedBy = $array[YiiAnnotationModel::MOTIVATED_BY];
+        $this->targets = $array[YiiAnnotationModel::TARGETS];
+    }
+
+    /**
+     * @return array contenant l'élément à enregistrer en base de données
```
[the remark this answers is not in the document]
Comment (niio972) [reply models/yiiModels/YiiAnnotationModel.php]: Done

File: models/yiiModels/YiiAnnotationModel.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            YiiAnnotationModel::URI => YiiAnnotationModel::URI_LABEL,
+            YiiAnnotationModel::CREATOR => Yii::t('app', YiiAnnotationModel::CREATOR_LABEL),
+            YiiAnnotationModel::MOTIVATED_BY => Yii::t('app', YiiAnnotationModel::MOTIVATED_BY_LABEL),
+            YiiAnnotationModel::COMMENTS => Yii::t('app', YiiAnnotationModel::COMMENTS_LABEL),
+            YiiAnnotationModel::TARGETS => Yii::t('app', YiiAnnotationModel::TARGETS_LABEL)
+        ];
+    }
+
+    /**
+     * Permet de remplir les attributs en fonction des informations
```
[the remark this answers is not in the document]
Comment (niio972) [reply models/yiiModels/YiiAnnotationModel.php]: Done

File: models/yiiModels/YiiAnnotationModel.php [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        $this->page = ($page !== null || $pageSize === "") ? $page : null;
+    }
+
+    /**
+     * 
+     * @inheritdoc
+     */
+    public function rules() {
+        return [
+            [[YiiAnnotationModel::URI, YiiAnnotationModel::CREATOR, YiiAnnotationModel::MOTIVATED_BY, YiiAnnotationModel::COMMENTS, YiiAnnotationModel::TARGETS], 'required'],
```
[the remark this answers is not in the document]
Comment (niio972) [reply models/yiiModels/YiiAnnotationModel.php]: Me too :+1:

File: models/yiiModels/YiiAnnotationModel.php [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+ * Implements a customized Active Record
+ *  (WSActiveRecord, for the web services access)
+ * @see app\models\wsModels\WSAnnotationModel
+ * @see app\models\wsModels\WSActiveRecord
+ * @author Morgane Vidal <morgane.vidal@inra.fr> 
+ * @author Arnaud Charleroy <arnaud.charleroy@.fr>
+ */
+class YiiAnnotationModel extends WSActiveRecord {
+
+    public $label = "Annotation";
```
[the remark this answers is not in the document]
Comment (niio972) [reply models/yiiModels/YiiAnnotationModel.php]: Thanks you for the reply, I changed it.

File: models/wsModels/WSAnnotationModel.php [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+include_once '../config/web_services.php';
+
+/**
+ * Encapsulate the access to the annotations service
+ * @see \openSILEX\guzzleClientPHP\WSModel
+ * @author Morgane Vidal <morgane.vidal@inra.fr>
+ * @author Arnaud Charleroy <arnaud.charleroy@inra.fr>
+ */
+class WSAnnotationModel extends \openSILEX\guzzleClientPHP\WSModel {
+    public static $ANNOTATION_SERVICE_NAME = "annotations";
```
[the remark this answers is not in the document]
Comment (niio972) [reply models/wsModels/WSAnnotationModel.php]: Not really, I changed it.

File: controllers/AnnotationController.php [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        if ($annotationModel->load(Yii::$app->request->post())) {
+            // Set model creator 
+            $userModel = new YiiUserModel();
+            $userModel->findByEmail($sessionToken, Yii::$app->session['email']);
+            $annotationModel->creator = $userModel->uri;
+            $annotationModel->isNewRecord = true;
+            $dataToSend[] = $annotationModel->attributesToArray();
+            // send data
+            $requestRes = $annotationModel->insert($sessionToken, $dataToSend);
+            if (is_string($requestRes) && $requestRes === \app\models\wsModels\WSConstants::TOKEN) { //L'utilisateur doit se connecter
```
[the remark this answers is not in the document]
Comment (niio972) [reply controllers/AnnotationController.php]: Done

File: components/helpers/Vocabulary.php [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+//                                       Vocabulary.php
+//
+// Author(s): Arnaud Charleroy <arnaud.charleroy@inra.fr>
+// PHIS-SILEX version 1.0
+// Copyright © - INRA - 2018
+// Creation date: 13 july 2018
+// Contact: arnaud.charleroy@inra.fr, anne.tireau@inra.fr, pascal.neveu@inra.fr
+// Last modification date:  13 july 2018
+// Subject: A helper used to format RDF
+//******************************************************************************
```
Comment (annetireau) [inline components/helpers/Vocabulary.php]: Please use the new coding style
